plugins/gitlab/api.ml

```diff
@@ -0,0 +1,644 @@
+open Lwt.Infix
+open Current.Syntax
+
+(* Limit updates to one at a time for now. *)
+let pool = Current.Pool.create ~label:"gitlab" 1
+
+(* When we get a webhook event, we fire the condition for that repository `owner/name`, if any. *)
+let webhook_cond = Hashtbl.create 10
+
+exception Project_not_found of int
+exception Commit_not_found
+
+type webhooks_accepted =
+  [ `MergeRequest of Gitlab_t.merge_request_webhook
+  | `Push of Gitlab_t.push_webhook
+  ]
+
+let input_webhook (body : webhooks_accepted) =
+  let update owner_name = match Hashtbl.find_opt webhook_cond owner_name with
+      | Some cond -> Lwt_condition.broadcast cond ()
+      | None -> Log.info (fun f -> f "Got webhook event for %S, but we're not interested in that" owner_name)
+  in
+  match body with
+  | `MergeRequest mr ->
+    update mr.merge_request_webhook_project.project_webhook_path_with_namespace
+  | `Push p ->
+    update  p.push_webhook_project.project_webhook_path_with_namespace
+
+let read_file path =
+  let ch = open_in_bin path in
+  Fun.protect
+    (fun () ->
+       let len = in_channel_length ch in
+       really_input_string ch len
+    )
+    ~finally:(fun () -> close_in ch)
+
+(* TODO Rework how we get a token. *)
+type token = {
+  token : (string, [`Msg of string]) result;
+  expiry : float option;
+}
+
+let no_token = {
+  token = Error (`Msg "Not fetched yet");
+  expiry = Some (-1.0);
+}
+
+module Repo_map = Map.Make(Repo_id)
+
+module Status = struct
+  type state = [`Cancelled | `Failure | `Running | `Pending | `Success ]
+
+  type t = {
+    state: state;
+    name: string;
+    description: string option;
+    url : Uri.t option;
+  }
+
+  let v ~name ?description ?url state =
+    { state; name; description; url }
+
+  let state_to_string = function
+    | `Cancelled -> "cancelled"
+    | `Failure -> "failure"
+    | `Running -> "running"
+    | `Pending -> "pending"
+    | `Success -> "success"
+
+  (* TODO JSON version of this might be unnecessary! *)
+  let json_items { state; name; description; url } =
+    ["state", `String (state_to_string state)
+    ; "name", `String name] @
+    (match description with None -> [] | Some x -> ["description", `String x]) @
+    (match url with None -> [] | Some x -> ["target_url", `String (Uri.to_string x)])
+
+  let digest t = Yojson.Safe.to_string @@ `Assoc (json_items t)
+
+  let pp f t = Fmt.string f (digest t)
+end
+
+module Ref = struct
+  type t = [ `Ref of string | `PR of int ] [@@deriving to_yojson]
+
+  let compare = Stdlib.compare
+
+  let pp f = function
+    | `Ref r -> Fmt.string f r
+    | `PR pr -> Fmt.pf f "PR %d" pr
+
+  let to_git = function
+    | `Ref head -> head
+    | `PR id -> Fmt.str "refs/pull/%d/head" id
```

Comment:
While testing, I've encountered failures because the ref to fetch is not correct

```suggestion
    | `PR id -> Fmt.str "refs/merge-requests/%d/head" id
```